### PR TITLE
[TOOLS-1257] Project Cluster specialization

### PR DESCRIPTION
**Pull request by JuniorMiqueletti**

### CODE REVIEW

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+import org.apache.commons.lang3.Validate;
+import org.hibernate.annotations.Check;
+
+import objective.taskboard.domain.TaskboardEntity;
+
+@Entity
+@Table(name = "followup_cluster_item")
+@Check(constraints = "(PROJECT_KEY IS NULL AND FOLLOWUP_CLUSTER_ID IS NOT NULL) OR (PROJECT_KEY IS NOT NULL AND FOLLOWUP_CLUSTER_ID IS NULL)")
+public class FollowUpClusterItemEntity extends TaskboardEntity {
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java]: MINOR Override the "equals" method in this class. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2160)

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterEntity.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.apache.commons.lang3.Validate;
+import org.hibernate.annotations.Fetch;
+import org.hibernate.annotations.FetchMode;
+
+import objective.taskboard.domain.TaskboardEntity;
+
+@Entity
+@Table(name = "followup_cluster")
+public class FollowUpClusterEntity extends TaskboardEntity {
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/followup/cluster/FollowUpClusterEntity.java]: MINOR Override the "equals" method in this class. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2160)

File: src/main/java/objective/taskboard/followup/cluster/FollowupClusterProvider.java [hunk 3]
```diff
@@ -14,20 +14,49 @@
...
 
-    public FollowupCluster getFor(ProjectFilterConfiguration project) {
-        List<FollowUpClusterItem> clusterItems = clusterItemRepository.findByProject(project);
-        if(isEmpty(clusterItems))
+    public FollowupCluster getFor(final ProjectFilterConfiguration project) {
+        final Optional<Long> followupClusterId = project.getFollowupClusterId();
+        final Optional<List<FollowUpClusterItemEntity>> items;
+        
+        if(followupClusterId.isPresent()) {
+            FollowUpClusterEntity cluster = clusterRepository.findById(project.getFollowupClusterId().get());
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/followup/cluster/FollowupClusterProvider.java]: MAJOR Call "Optional#isPresent()" before accessing the value. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3655)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/resources/db/migration/mysql/V201807251042__projectClusterSpecialization.sql [hunk 4]
```diff
@@ -0,0 +1,50 @@
+-- Create Followup cluster table
+
+CREATE TABLE followup_cluster
```
Comment (taksan) [inline src/main/resources/db/migration/mysql/V201807251042__projectClusterSpecialization.sql]: sizing_cluster

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    public Double getCycle() {
+        return cycle;
+    }
+
+    public void setCycle(Double cycle) {
+        this.cycle = cycle;
+    }
+
+    public FollowUpClusterEntity getFollowupCluster() {
```
Comment (taksan) [inline src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java]: Optional<...>

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        return followUpCluster;
+    }
+
+    public void setFollowupCluster(FollowUpClusterEntity followupClusterEntity) {
+        this.followUpCluster = followupClusterEntity;
+    }
+
+    public String getProjectKey() {
```
Comment (taksan) [inline src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java]: Optional<...>

File: src/main/java/objective/taskboard/followup/cluster/FollowupClusterProvider.java [hunk 7]
```diff
@@ -14,20 +14,49 @@
...
+            items = clusterItemRepository.findByProjectKeyOrFollowUpCluster(project.getProjectKey(), cluster);
+            
+        } else {
+            items = clusterItemRepository.findByProjectKey(project.getProjectKey());
+        }
+        
+        if(!items.isPresent())
             return new EmptyFollowupCluster();
+        
+        List<FollowUpClusterItem> clusterItems =
```
Comment (taksan) [inline src/main/java/objective/taskboard/followup/cluster/FollowupClusterProvider.java]: Looks like we have a problem here. What about the overrides?

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java (hunk 1 above)
Comment (taksan) [reply src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItemEntity.java]: SizingClusterItem

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterEntity.java (hunk 2 above)
Comment (taksan) [inline src/main/java/objective/taskboard/followup/cluster/FollowUpClusterEntity.java]: Remove the Entity suffix; Let's call it BaseSizingCluster

File: src/main/java/objective/taskboard/domain/ProjectFilterConfiguration.java [hunk 8]
```diff
@@ -145,4 +148,12 @@ public void setBaselineDate(LocalDate baselineDate) {
     public Optional<LocalDate> getBaselineDate() {
         return Optional.ofNullable(baselineDate);
     }
+
+    public Optional<Long> getFollowupClusterId() {
```
Comment (taksan) [inline src/main/java/objective/taskboard/domain/ProjectFilterConfiguration.java]: Not optional. Let's force a cluster

File: src/main/java/objective/taskboard/followup/cluster/FollowupClusterProvider.java [hunk 9]
```diff
@@ -14,20 +17,62 @@
...
+
+        allItems.stream()
+            .filter(i-> i.getProjectKey().isPresent())
+            .sorted(Comparator.comparing(SizingClusterItem::getId).reversed())
+            .forEach(i -> clusterItemsMap.put(i.getSubtaskTypeName() + i.getParentTypeName() + i.getSizing(),
+                                                convertToFollowUpClusterItem(project, i)));
+
+        return clusterItemsMap.entrySet().parallelStream()
+                    .map(Entry<String, FollowUpClusterItem>::getValue)
+                    .sorted(Comparator.comparing(FollowUpClusterItem::getSubtaskTypeName))
```
Comment (taksan) [inline src/main/java/objective/taskboard/followup/cluster/FollowupClusterProvider.java]: Should compare the whole key
i.getSubtaskTypeName() + i.getParentTypeName() + i.getSizing()

File: src/main/resources/db/migration/mysql/V201807251042__projectClusterSpecialization.sql [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ALTER TABLE project_filter_configuration
+    ADD CONSTRAINT project_cluster_fk
+    FOREIGN KEY (base_cluster_id) REFERENCES sizing_cluster(id);
+
+ALTER TABLE project_filter_configuration
+    MODIFY base_cluster_id bigint(20) NOT NULL;
+
+-- followup_cluster_item adjustment
+
+ALTER TABLE followup_cluster_item
```
Comment (taksan) [inline src/main/resources/db/migration/mysql/V201807251042__projectClusterSpecialization.sql]: It's odd that you are changing the followup_cluster_item and after that you drop the table. Why not alter sizing_cluster_item instead?

File: src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItem.java [hunk 11]
```diff
@@ -1,38 +1,26 @@
...
-@Entity
-@Table(name = "followup_cluster_item")
-public class FollowUpClusterItem extends TaskboardEntity {
+public class FollowUpClusterItem {
 
-    @ManyToOne(optional = false, fetch = FetchType.EAGER)
-    @JoinColumn(name = "project_key", referencedColumnName = "projectKey", nullable = false)
     private ProjectFilterConfiguration project;
-
     private String subtaskTypeName;
```
Comment (taksan) [inline src/main/java/objective/taskboard/followup/cluster/FollowUpClusterItem.java]: Make all fields final if you want hashcode implemented

File: src/test/java/objective/taskboard/followup/FollowUpClusterProviderTest.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        when(clusterItemRepository.findByProjectKeyOrBaseCluster(TASKB_PROJECT_KEY, baseCluster))
+            .thenReturn(Optional.of(itemsByProjectMock));
+        
+        when(clusterItemRepository.findByProjectKeyOrBaseCluster(TASKB_THREE_PROJECT_KEY, baseCluster))
+            .thenReturn(Optional.of(itemsProjectThreeMock));
+    }
+
+    @Test
+    public void shouldReturnEmptyWhen_projectNotFound() {
+        FollowupCluster followupCluster = followupClusterProvider.getForProject(UNKNOWN_PROJECT_KEY);
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/FollowUpClusterProviderTest.java]: Move the specific mock setups to the tests

File: src/test/java/objective/taskboard/followup/FollowUpClusterProviderTest.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        List<FollowUpClusterItem> expectedItems = Arrays.asList(
+                new FollowUpClusterItem(expectedProject, BALLPARK_ALPHA, NOTUSED, "XS", 1.0, 0.0),
+                new FollowUpClusterItem(expectedProject, BALLPARK_DEV, NOTUSED, "M", 5.0, 0.0),
+                new FollowUpClusterItem(expectedProject, DEV, NOTUSED, "S", 2.0, 0.0)
+                );
+        
+        assertFollowUpItemsEquals(expectedItems, clusterItems);
+    }
+    
+    private void assertFollowUpItemsEquals(List<FollowUpClusterItem> expectedItems,
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/FollowUpClusterProviderTest.java]: This method would be better at the bottom of the class.

Comment (taksan) [reply src/test/java/objective/taskboard/followup/FollowUpClusterProviderTest.java]: Tip: if you make in a way that each item shows in a line, if the test fails, the comparison will be nicer
assertEquals(expectedItems.join("\n"), clusterItems.join("\n"))

**Pull request merged**